static/js/tools/choropleth/choropleth.tsx

```diff
@@ -0,0 +1,351 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Creates and manages choropleth rendering.
+ */
+
+import React, { Component } from "react";
+import axios from "axios";
+import * as d3 from "d3";
+
+class ChoroplethMap extends Component {
+    constructor(props: any) {
+        super(props);
+
+        // Redirect to basic url if none provided.
+        if (window.location.search === "") {
+            window.location.search = "statVar=Count_Person_Employed&perCapita=t&geoId=country/USA";
+        }
+        
+        // Add default state and bind function contexts.
+        const urlParams = new URLSearchParams(window.location.search);
+        this.state = {
+            geojson: [],
+            perCapita: (urlParams.has("perCapita")
+                && (["t","true", "1"]).includes(
+                    urlParams.get("perCapita").toLowerCase())),
+            values: {},
+        };
+        this.loadGeoJson = this.loadGeoJson.bind(this);
+        this.renderGeoMap = this.renderGeoMap.bind(this);
+        this.loadValues = this.loadValues.bind(this);
+        this.handleMapHover = this.handleMapHover.bind(this);
+        this.updateGeoValues = this.updateGeoValues.bind(this);        
+        this.loadGeoJson();
+    }
+
+    /**
+     * Loads and renders blank GeoJson map for current geoId.
+     * After loading, values for a particular StatVar are pulled.
+     */
+    loadGeoJson() {
+        const urlParams = new URLSearchParams(window.location.search);
+        const baseUrl = buildChoroplethUrl(
+            ["perCapita", "geoId", "level", "mdom"],
+            "/geo",
+        );
+
+        // Create request and generate map.
+        axios.get(baseUrl).then((resp) => {
+            this.setState({"geojson": resp.data[0]});
+            this.renderGeoMap();
+            this.loadValues();
+        });  
+    }
+
+    /**
+     * Loads values for the current geoId and updates map.
+     */
+    loadValues() {
+        const urlParams = new URLSearchParams(window.location.search);
+        const baseUrl = buildChoroplethUrl(
+            ["geoId", "statVar"],
+            "/values",
+        );
+
+        axios.get(baseUrl).then((resp) => {
+            this.setState({"values": resp.data[0]});
+            this.updateGeoValues();
+        });
```

Comment:
Can you fetch data from both endpoints in parallel using Promises.all()

Reply:
Similarly, updateGeoValues should be in shouldComponentUpdate

Reply:
Wrapped